File: c/complex-numbers/complex_numbers.c
#include "complex_numbers.h"
#include <math.h>

complex_t c_add(complex_t a, complex_t b) {
  return (complex_t){ a.real + b.real, a.imag + b.imag };
}

complex_t c_sub(complex_t a, complex_t b) {
  return (complex_t){ a.real - b.real, a.imag - b.imag };
}

complex_t c_mul(complex_t a, complex_t b) {
  return (complex_t){ a.real * b.real - a.imag * b.imag,
                      a.imag * b.real + a.real * b.imag };
}

complex_t c_div(complex_t a, complex_t b) {
  return (complex_t){ (a.real * b.real + a.imag * b.imag) / 
                        (b.real * b.real + b.imag * b.imag),
                       (a.imag * b.real - a.real * b.imag) /
                        (b.real * b.real + b.imag * b.imag) };
}

double c_abs(complex_t x) {
  return pow(x.real * x.real + x.imag * x.imag, 0.5);
}

complex_t c_conjugate(complex_t x) {
  return (complex_t){x.real, -x.imag};
}

double c_real(complex_t x) {
  return x.real;
}

double c_imag(complex_t x) {
  return x.imag;
}

complex_t c_exp(complex_t x) {
  return c_mul((complex_t){ exp(x.real), 0}, (complex_t) {cos(x.imag), sin(x.imag)});
}

File: javascript/complex-numbers/complex-numbers.js
export class ComplexNumber {
  constructor(real = 0, imag = 0) {
    this._real = real;
    this._imag = imag;
  }

  get real() { return this._real; }

  get imag() { return this._imag; }

  add(cn) {
    return new ComplexNumber(this.real + cn.real, this.imag + cn.imag);
  }

  sub(cn) {
    return new ComplexNumber(this.real - cn.real, this.imag - cn.imag);
  }

  div(cn) {
    const _real = (this.real * cn.real + this.imag * cn.imag) /
                 (cn.real ** 2 + cn.imag ** 2);
    const _imag = (this.imag * cn.real - this.real * cn.imag) /
                 (cn.real ** 2 + cn.imag ** 2);
    return new ComplexNumber(_real, _imag);
  }

  mul(cn) {
    const _real = (this.real * cn.real - this.imag * cn.imag);
    const _imag = (this.imag * cn.real + this.real * cn.imag);
    return new ComplexNumber(_real, _imag);
  }

  get abs() {
    return (this.real ** 2 + this.imag ** 2) ** 0.5;
  }

  get conj() {
    return new ComplexNumber(this.real, 0 - this.imag);
  }

  get exp() {
    return new ComplexNumber(Math.exp(this.real), 0)
      .mul(new ComplexNumber(Math.cos(this.imag), Math.sin(this.imag)));
  }
}
